src/robot/running/keywordrunner.py

```diff
@@ -206,16 +256,99 @@ def _to_number_with_arithmetics(self, item):
+                % (len(data.values), len(data.variables)))
+
+    def _transform_items(self, items):
+        answer = list()
+        for item in items:
+            if not is_list_like(item):
+                # Checking here because _validate() doesn't have access
+                # to the actual values.
+                raise DataError("For-In-Zip Loop items must all be List-like (but not Strings); got %s with value '%s'" % (item.__class__, item))
+        for zipped_item in zip(*[list(item) for item in items]):
+            answer.extend(zipped_item)
+        return answer
+
+
+class ForInEnumerateLoopRunner(BasicForLoopRunner):
+    def __init__(self, context, templated=False):
+        super(ForInEnumerateLoopRunner, self).__init__(context, templated)
+
+    def _flavor_name(self):
+        return 'IN ENUMERATE'
+
+    def _original_values_per_iteration(self, values):
+        """
+        The number of values per iteration;
+        used to check if we have (a multiple of this) values.
+
+        This is its own method to support loops like ForInEnumerate
+        which add/remove items to the pool.
+        """
+        if len(values) < 2:
+            raise DataError('FOR IN ENUMERATE expected 2 or more loop variables, got %d.'
+                    % len(values))
+        return len(values) - 1
+
+    def _get_values_for_one_round(self, data):
+        for idx, values in enumerate(super(ForInEnumerateLoopRunner, self)
+                ._get_values_for_one_round(data)):
+            yield [idx] + values
+
+
+# Defining this dict here in for potentially easier extension by libraries...
+# ...unless that's a terrible idea?
+def ForLoopRunner(context, templated=False, flavor='IN'):
+    FOR_LOOP_FLAVORS = dict(
+            IN=BasicForLoopRunner,
+            INRANGE=ForInRangeLoopRunner,
+            INZIP=ForInZipLoopRunner,
+            INENUMERATE=ForInEnumerateLoopRunner,
+            )
+    flavor_key = flavor.upper().replace(' ', '')
+    if flavor_key in FOR_LOOP_FLAVORS:
+        runner_class = FOR_LOOP_FLAVORS[flavor_key]
+        return runner_class(context, templated)
+    else:
+        return InvalidForLoopRunner(context, flavor_key, flavor, FOR_LOOP_FLAVORS.keys())
+
+
+class InvalidForLoopRunner(BasicForLoopRunner):
+    """
+    Used to send an error from ForLoopRunner() if it sees an unexpected error.
+
+    We can't simply throw a DataError from ForLoopRunner() because that happens
+    outside the "with StatusReporter(...)" blocks.
+
+    """
+    def __init__(self, context, flavor, pretty_flavor, expected):
+        super(InvalidForLoopRunner, self).__init__(context, False)
+        self.flavor = flavor
+        self.pretty_flavor = pretty_flavor
+        self.expected = repr(sorted(expected))
+
+    def _run(self, data, *args, **kwargs):
+        raise DataError(
+            "Unexpected For-loop type %s (%s); expected one of %s (but with spaces)" %
+            (self.flavor, self.pretty_flavor, self.expected))
```

Comment:
I like dynamic code but here I'd just hard-code the types to make the error better. For example:

    "Invalid FOR loop type '%s'. Expected 'IN', 'IN RANGE', 'IN ZIP', or 'IN ENUMERATE'." % self.flavor